galley/pkg/server/process/host.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package process
+
+import (
+	"errors"
+	"sync"
+)
+
+// Host is a process host that takes care of start/stop of sub-components.
+type Host struct {
+	mu         sync.Mutex
+	started    bool
+	components []Component
+}
+
+// Add a new component to this host. It panics if the host is already started.
+func (h *Host) Add(c Component) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.started {
+		panic("Host.Add: host is already started")
+	}
+	h.components = append(h.components, c)
+}
+
+// Start the process.Host. If any of the components starts to fail, then an error is returned. If there is an error
+// then all successfully started components will be stopped before control is returned.
+func (h *Host) Start() error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.started {
+		return errors.New("process.Host: already started")
+	}
+
+	for i := 0; i < len(h.components); i++ {
+		c := h.components[i]
+		if err := c.Start(); err != nil {
+			// Component startup failed. Stop already started components.
```

Comment:
Why not use a context here? Is it just too heavyweight for what's actually needed in this case?

Reply:
you mean context.Context? The main usage model for that is RPC-like per requests, I believe. One issue with such models (i.e. using channels etc.) it is very hard to ensure that there is no background operation that is still active after closing the channel. Using explicit method calls help with this, as return from a method call can be a guarantee point for ensuring all background activity has ceased.